### ISSUE TITLE
Synthesized delegates allow ref safety holes on the boundary of old and updated rules

### ISSUE THREAD

**Issue by jjonescz** [dotnet/roslyn#76087]
```cs
[Fact]
public void RefSafetyRules_SynthesizedDelegate()
{
    var source1 = """
        public static class C
        {
            public static R M(ref int x) => new R { F = ref x };
        }
        public ref struct R
        {
            public ref int F;
        }
        """;
    var ref1 = CreateCompilation(source1, targetFramework: TargetFramework.Net70)
        .VerifyDiagnostics().EmitToImageReference();

    var source2 = """
        R r;
        {
            int x = 1;
            r = C.M(ref x);

            var d = C.M;
            r = d(ref x);
        }
        """;
    CreateCompilation([source2, UnscopedRefAttributeDefinition], [ref1],
        parseOptions: TestOptions.Regular10).VerifyDiagnostics(
        // (4,9): error CS8347: Cannot use a result of 'C.M(ref int)' in this context because it may expose variables referenced by parameter 'x' outside of their declaration scope
        //     r = C.M(ref x);
        Diagnostic(ErrorCode.ERR_EscapeCall, "C.M(ref x)").WithArguments("C.M(ref int)", "x").WithLocation(4, 9),
        // (4,17): error CS8168: Cannot return local 'x' by reference because it is not a ref local
        //     r = C.M(ref x);
        Diagnostic(ErrorCode.ERR_RefReturnLocal, "x").WithArguments("x").WithLocation(4, 17));
}
```

There is no error reported at `r = d(ref x);` even though the call is equivalent to `r = C.M(ref x);` where a ref safety error is reported.

**Comment by RikkiGibson** [comment]
I think the most reasonable solution here would be to make the delegate conversion an error, when a difference between the method and delegate signatures is violating ref safety.

Basically I think if there is some "old rules" method involving refs and ref structs, which user wants to make a delegate for in a newer project, they may have to do some contortions for the delegate type they declare, or enter an unsafe context and suppress a warning to proceed.

**Comment by jjonescz** [comment]
It looks like this is not limited to delegates. There are other ref safety holes on boundary of old and updated rules. Consider:

```cs
[Fact]
public void RefSafetyRules_Boundary()
{
    var source1 = """
        public static class C
        {
            public static R M(ref int x) => new R { F = ref x };
            public static int GetF(ref R r) => r.F;
        }
        public ref struct R
        {
            public ref int F;
        }
        """;
    var ref1 = CreateCompilation(source1, targetFramework: TargetFramework.Net70)
        .VerifyDiagnostics().EmitToImageReference();

    var source2 = """
        R r = F();
        System.Console.WriteLine(C.GetF(ref r)); // use after free

        static R M2(ref int y) => C.M(ref y); // "promoting" new-rules method to old-rules method; should not be allowed

        static R F()
        {
            int i = 2 * 2;
            return M2(ref i); // ok under old rules
        }
        """;

    CreateCompilation(source2, [ref1], targetFramework: TargetFramework.Net70).VerifyDiagnostics(
        // (9,19): error CS8168: Cannot return local 'i' by reference because it is not a ref local
        //     return M2(ref i);
        Diagnostic(ErrorCode.ERR_RefReturnLocal, "i").WithArguments("i").WithLocation(9, 19),
        // (9,12): error CS8347: Cannot use a result of 'M2(ref int)' in this context because it may expose variables referenced by parameter 'y' outside of their declaration scope
        //     return M2(ref i);
        Diagnostic(ErrorCode.ERR_EscapeCall, "M2(ref i)").WithArguments("M2(ref int)", "y").WithLocation(9, 12));

    CompileAndVerify(source2, [ref1],
        verify: Verification.Skipped,
        expectedOutput: "4",
        parseOptions: TestOptions.Regular10).VerifyDiagnostics();
}
```

**Comment by jjonescz** [comment]
Given this would require a larger fix and likely a spec update; would be a breaking change; affects only edge cases (boundary of old/new rules); and wasn't reported by a user; moving to Backlog.
